mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java

```diff
@@ -165,14 +164,30 @@ public Map<String, String> getSchedulingAttributes() {
 
     @JsonIgnore
     public TaskExecutorGroupKey getTaskExecutorGroupKey() {
-        return new TaskExecutorGroupKey(machineDefinition, getAttributeByKey(MANTIS_CONTAINER_SIZE_NAME_KEY), getSchedulingAttributes());
+        Optional<String> sizeName = getAttributeByKey(MANTIS_CONTAINER_SIZE_NAME_KEY)
+            .filter(name -> !name.matches("\\$\\{.*\\}"));
+        return new TaskExecutorGroupKey(machineDefinition, sizeName, getSchedulingAttributes());
     }
 
     @Value
-    @AllArgsConstructor
     public static class TaskExecutorGroupKey {
         MachineDefinition machineDefinition;
         Optional<String> sizeName;
         Map<String, String> schedulingAttributes;
+
+        public TaskExecutorGroupKey(MachineDefinition machineDefinition,
+                                    Optional<String> sizeName,
+                                    Map<String, String> schedulingAttributes) {
+            this.machineDefinition = new MachineDefinition(
+                Math.round(machineDefinition.getCpuCores()),
+                Math.round(machineDefinition.getMemoryMB()),
+                Math.round(machineDefinition.getNetworkMbps()),
+                Math.round(machineDefinition.getDiskMB()),
+                // TODO: remove this value.
+                machineDefinition.getNumPorts());
```

Comment:
Is this needed after your new set of changes?

Reply:
are you referring to the `Math.round`? If that's the case, the purpose is to safeguard the new groupings from "minimal double fluctuations" (which in practice are rare but they could still happen). Utilizing integers could potentially minimize grouping TE errors.